toolchain/diagnostics/diagnostic_emitter.h

```diff
@@ -295,8 +306,37 @@ class DiagnosticEmitter {
   }
 
  private:
+  // Base class for scopes in which we perform diagnostic annotation, such as
+  // adding notes with contextual information.
+  class DiagnosticAnnotationScopeBase {
```

Comment:
Why the separation of logic here, versus something like `llvm::SmallVector<std::function<void(DiagnosticBuilder)>>`?

Maybe this thought is a little too quirky/limiting, but:

```
// Returns a scope_exit, so we let the return type be inferred.
auto AddAnnotationScope(std::function<void(DiagnosticBuilder)> annotator) {
  annotators_.push_back(annotator);
  return llvm::make_scope_exit([annotator]() {
    CARBON_CHECK(emitter_->annotators_.back() == annotator);
    annotators.pop_back();
  });
}
```

Reply:
This is efficiency-driven; `std::function` will perform a heap allocation if the lambda doesn't fit into the small-function buffer, and often performs two indirect function calls to invoke its callee rather than one. (Also, minor, `std::function` doesn't have an `operator==` so we can't perform the `CHECK`.)

The `AddAnnotationScope` approach is definitely interesting. I'm going to land this first and then try that out.

Reply:
Bah, I missed that https://en.cppreference.com/w/cpp/utility/functional/function says `==` was removed in c++20